artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java

```diff
@@ -2920,6 +2922,10 @@ private boolean deliver() {
       // Either the iterator is empty or the consumer is busy
       int noDelivery = 0;
 
+      // track filters not matching, used to track when all consumers can't match, redistribution is then an option
+      int numNoMatch = 0;
+      final int consumerCount = this.consumers.size();
```

Comment:
Between here and the sync block a consumer could be added or removed and thus the count and logic within sync block would be broke. This probably should be in the sync block to avoid this

Reply:
Or just call consumers.size when needed to be safe. Rather than causing race issues